Be able to KILL scalar queries

Algunenano opened the pull request:
Changelog category (leave one):
- Backward Incompatible Change


Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Account for scalar subqueries. With this change, rows read in scalar subqueries are now reported in the query_log. If the scalar subquery is cached (repeated or called for several rows) the rows read are only counted once. This change allows KILLing queries and reporting progress while they are executing scalar subqueries.




Detailed description / Documentation draft:

As https://github.com/ClickHouse/ClickHouse/pull/31636 added checks in the pipelines, add the necessary process list elements to pipelines created with `QueryPlan::buildQueryPipeline` so that they too can be KILLed, handle timeouts and so on.

This partially covers https://github.com/ClickHouse/ClickHouse/issues/1576. With this change the queries can be killed, but Ctrl+C in the client still doesn't work because the TCPHandler is stuck in executeQuery and doesn't process network packets.

By fixing this it turned out to also improve progress reports:
* Scalar subqueries report it's progress, events, etc. For example `Select * from table where x in (Select * from subquery)` will now also include in its metrics the rows read in subquery (before they were considered 0).
* Inserts with materialized views dependencies now report rows read by the dependent inserts as part of the original query. Before it only reported rows written (but not read) in the dependent queries.
* It also starts using `progress_out.written_rows` instead of `progress_out.read_rows`. It makes more sense and there was a few TODOs related to it.

Closes #32993

Algunenano committed:
Force PipeLineExecutor creators to pass a QueryStatus

This is to avoid future issues

Algunenano committed:
Add QueryStatus to BuildQueryPipelineSettings and use it

Algunenano committed:
Test that scalar queries can be killed

KochetovNicolai commented:
However, there is a crash in tests
```
2021.12.06 15:40:08.690148 [ 22491 ] {} <Fatal> BaseDaemon: ########################################
2021.12.06 15:40:08.690169 [ 22491 ] {} <Fatal> BaseDaemon: (version 21.12.1.1, build id: A4EB42B6BD56438C) (from thread 18319) (query_id: 1fa7eb2a-94d7-45d0-9fa7-eb2a94d795d0::all_0_1_1_2) Received signal Segmentation fault (11)
2021.12.06 15:40:08.690185 [ 22491 ] {} <Fatal> BaseDaemon: Address: 0x218 Access: write. Address not mapped to object.
2021.12.06 15:40:08.690195 [ 22491 ] {} <Fatal> BaseDaemon: Stack trace: 0x5d86d63 0xd787288 0xc73d5d3 0xc74498c 0xd4ef168 0xd4ee4dd 0xd4e276e 0xd504975 0xd503de2 0xd3ac69a 0xd3ace39 0x5d7efa9 0x5d8064c 0x5d7d54a 0x5d7f92e 0x7f593dcb9609 0x7f593df44293
2021.12.06 15:40:08.691742 [ 22491 ] {} <Fatal> BaseDaemon: 3.1. inlined from /fasttest-workspace/build/../../ClickHouse/contrib/libcxx/include/atomic:1050: unsigned long std::__1::__cxx_atomic_fetch_add<unsigned long>(std::__1::__cxx_atomic_base_impl<unsigned long>*, unsigned long, std::__1::memory_order)
2021.12.06 15:40:08.691755 [ 22491 ] {} <Fatal> BaseDaemon: 3.2. inlined from ../../ClickHouse/contrib/libcxx/include/atomic:1719: std::__1::__atomic_base<unsigned long, true>::fetch_add(unsigned long, std::__1::memory_order)
2021.12.06 15:40:08.691762 [ 22491 ] {} <Fatal> BaseDaemon: 3.3. inlined from ../../ClickHouse/contrib/libcxx/include/atomic:1764: std::__1::__atomic_base<unsigned long, true>::operator+=(unsigned long)
2021.12.06 15:40:08.691768 [ 22491 ] {} <Fatal> BaseDaemon: 3. ../../ClickHouse/src/IO/Progress.cpp:69: DB::Progress::incrementPiecewiseAtomically(DB::Progress const&) @ 0x5d86d63 in /fasttest-workspace/build/programs/clickhouse
2021.12.06 15:40:08.696909 [ 22491 ] {} <Fatal> BaseDaemon: 4.1. inlined from /fasttest-workspace/build/../../ClickHouse/contrib/libcxx/include/memory:2835: std::__1::shared_ptr<DB::QueryPriorities::HandleImpl>::get() const
2021.12.06 15:40:08.696924 [ 22491 ] {} <Fatal> BaseDaemon: 4.2. inlined from ../../ClickHouse/contrib/libcxx/include/memory:2851: std::__1::shared_ptr<DB::QueryPriorities::HandleImpl>::operator bool() const
2021.12.06 15:40:08.696932 [ 22491 ] {} <Fatal> BaseDaemon: 4.3. inlined from ../../ClickHouse/src/Interpreters/ProcessList.h:153: DB::QueryStatus::updateProgressIn(DB::Progress const&)
2021.12.06 15:40:08.696938 [ 22491 ] {} <Fatal> BaseDaemon: 4. ../../ClickHouse/src/Processors/Sources/SourceWithProgress.cpp:44: DB::SourceWithProgress::setProcessListElement(DB::QueryStatus*) @ 0xd787288 in /fasttest-workspace/build/programs/clickhouse
2021.12.06 15:40:08.709591 [ 22491 ] {} <Fatal> BaseDaemon: 5. /fasttest-workspace/build/../../ClickHouse/src/QueryPipeline/QueryPipeline.cpp:0: DB::QueryPipeline::setProcessListElement(DB::QueryStatus*) @ 0xc73d5d3 in /fasttest-workspace/build/programs/clickhouse
2021.12.06 15:40:08.725098 [ 22491 ] {} <Fatal> BaseDaemon: 6. /fasttest-workspace/build/../../ClickHouse/src/QueryPipeline/QueryPipelineBuilder.cpp:565: DB::QueryPipelineBuilder::getPipeline(DB::QueryPipelineBuilder) @ 0xc74498c in /fasttest-workspace/build/programs/clickhouse
2021.12.06 15:40:08.779425 [ 22491 ] {} <Fatal> BaseDaemon: 7. /fasttest-workspace/build/../../ClickHouse/src/Storages/MergeTree/MutateTask.cpp:0: DB::MutateAllPartColumnsTask::prepare() @ 0xd4ef168 in /fasttest-workspace/build/programs/clickhouse
2021.12.06 15:40:08.834577 [ 22491 ] {} <Fatal> BaseDaemon: 8. /fasttest-workspace/build/../../ClickHouse/src/Storages/MergeTree/MutateTask.cpp:0: DB::MutateAllPartColumnsTask::executeStep() @ 0xd4ee4dd in /fasttest-workspace/build/programs/clickhouse
2021.12.06 15:40:08.873787 [ 22491 ] {} <Fatal> BaseDaemon: 9. /fasttest-workspace/build/../../ClickHouse/src/Storages/MergeTree/MutateTask.cpp:1229: DB::MutateTask::execute() @ 0xd4e276e in /fasttest-workspace/build/programs/clickhouse
2021.12.06 15:40:08.878203 [ 22491 ] {} <Fatal> BaseDaemon: 10. /fasttest-workspace/build/../../ClickHouse/src/Storages/MergeTree/ReplicatedMergeMutateTaskBase.cpp:0: DB::ReplicatedMergeMutateTaskBase::executeImpl() @ 0xd504975 in /fasttest-workspace/build/programs/clickhouse
2021.12.06 15:40:08.882301 [ 22491 ] {} <Fatal> BaseDaemon: 11. /fasttest-workspace/build/../../ClickHouse/src/Storages/MergeTree/ReplicatedMergeMutateTaskBase.cpp:0: DB::ReplicatedMergeMutateTaskBase::executeStep() @ 0xd503de2 in /fasttest-workspace/build/programs/clickhouse
2021.12.06 15:40:08.886433 [ 22491 ] {} <Fatal> BaseDaemon: 12. /fasttest-workspace/build/../../ClickHouse/src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:95: DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(std::__1::shared_ptr<DB::TaskRuntimeData>) @ 0xd3ac69a in /fasttest-workspace/build/programs/clickhouse
2021.12.06 15:40:08.890715 [ 22491 ] {} <Fatal> BaseDaemon: 13.1. inlined from /fasttest-workspace/build/../../ClickHouse/contrib/libcxx/include/memory:3211: ~shared_ptr
2021.12.06 15:40:08.890727 [ 22491 ] {} <Fatal> BaseDaemon: 13. ../../ClickHouse/src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:177: DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::threadFunction() @ 0xd3ace39 in /fasttest-workspace/build/programs/clickhouse
2021.12.06 15:40:08.897362 [ 22491 ] {} <Fatal> BaseDaemon: 14.1. inlined from /fasttest-workspace/build/../../ClickHouse/contrib/libcxx/include/functional:2210: std::__1::__function::__policy_func<void ()>::operator=(std::nullptr_t)
2021.12.06 15:40:08.897374 [ 22491 ] {} <Fatal> BaseDaemon: 14.2. inlined from ../../ClickHouse/contrib/libcxx/include/functional:2533: std::__1::function<void ()>::operator=(std::nullptr_t)
2021.12.06 15:40:08.897380 [ 22491 ] {} <Fatal> BaseDaemon: 14. ../../ClickHouse/src/Common/ThreadPool.cpp:277: ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) @ 0x5d7efa9 in /fasttest-workspace/build/programs/clickhouse
2021.12.06 15:40:08.902578 [ 22491 ] {} <Fatal> BaseDaemon: 15. /fasttest-workspace/build/../../ClickHouse/src/Common/ThreadPool.cpp:0: ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()&&...)::'lambda'()::operator()() @ 0x5d8064c in /fasttest-workspace/build/programs/clickhouse
2021.12.06 15:40:08.908504 [ 22491 ] {} <Fatal> BaseDaemon: 16.1. inlined from /fasttest-workspace/build/../../ClickHouse/contrib/libcxx/include/functional:2210: std::__1::__function::__policy_func<void ()>::operator=(std::nullptr_t)
2021.12.06 15:40:08.908520 [ 22491 ] {} <Fatal> BaseDaemon: 16.2. inlined from ../../ClickHouse/contrib/libcxx/include/functional:2533: std::__1::function<void ()>::operator=(std::nullptr_t)
2021.12.06 15:40:08.908527 [ 22491 ] {} <Fatal> BaseDaemon: 16. ../../ClickHouse/src/Common/ThreadPool.cpp:277: ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x5d7d54a in /fasttest-workspace/build/programs/clickhouse
2021.12.06 15:40:08.915721 [ 22491 ] {} <Fatal> BaseDaemon: 17.1. inlined from /fasttest-workspace/build/../../ClickHouse/contrib/libcxx/include/memory:1655: std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >::reset(std::__1::__thread_struct*)
2021.12.06 15:40:08.915733 [ 22491 ] {} <Fatal> BaseDaemon: 17.2. inlined from ../../ClickHouse/contrib/libcxx/include/memory:1612: ~unique_ptr
2021.12.06 15:40:08.915739 [ 22491 ] {} <Fatal> BaseDaemon: 17.3. inlined from ../../ClickHouse/contrib/libcxx/include/tuple:181: ~__tuple_leaf
2021.12.06 15:40:08.915745 [ 22491 ] {} <Fatal> BaseDaemon: 17.4. inlined from ../../ClickHouse/contrib/libcxx/include/tuple:477: ~tuple
2021.12.06 15:40:08.915755 [ 22491 ] {} <Fatal> BaseDaemon: 17.5. inlined from ../../ClickHouse/contrib/libcxx/include/memory:1397: std::__1::default_delete<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()> >::operator()(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()>*) const
2021.12.06 15:40:08.915772 [ 22491 ] {} <Fatal> BaseDaemon: 17.6. inlined from ../../ClickHouse/contrib/libcxx/include/memory:1658: std::__1::unique_ptr<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()>, std::__1::default_delete<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()> > >::reset(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()>*)
2021.12.06 15:40:08.915778 [ 22491 ] {} <Fatal> BaseDaemon: 17.7. inlined from ../../ClickHouse/contrib/libcxx/include/memory:1612: ~unique_ptr
2021.12.06 15:40:08.915784 [ 22491 ] {} <Fatal> BaseDaemon: 17. ../../ClickHouse/contrib/libcxx/include/thread:293: void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()> >(void*) @ 0x5d7f92e in /fasttest-workspace/build/programs/clickhouse
2021.12.06 15:40:08.915792 [ 22491 ] {} <Fatal> BaseDaemon: 18. ? @ 0x7f593dcb9609 in ?
2021.12.06 15:40:08.915798 [ 22491 ] {} <Fatal> BaseDaemon: 19. clone @ 0x7f593df44293 in ?
```

Algunenano committed:
SourceWithProgress: Don't use process_list_elem without checking it

Algunenano commented:
The issue was that `SourceWithProgress::setProcessListElement` assumed that the passed pointer would not be null and used it without checking it, but with the changes that's no longer the case.

Algunenano commented:
Looks like progress (from `SourceWithProgress`) is now reported multiple times.

Algunenano commented:
This change has surfaced several issues that I can see:
* `QueryPipeline::setProcessListElement` has side effects because `SourceWithProgress::setProcessListElement` has side effects. It might make sense to not do anything if the new process list element is the same as the old.

* Some tests, for example [01583_const_column_in_set_index.sql](https://github.com/ClickHouse/ClickHouse/blob/6b4512809cc195b9d6c86789733468e175599910/tests/queries/0_stateless/01583_const_column_in_set_index.sql) seem to be relying on the fact that IN set subqueries progress wasn't being reported. Now that it is, things break. As far as I can see, in that specific test (line 7) you are reading 10 rows from numbers() + 2 rows from insub, for a total of 12 rows.

* The same happens, for example, in `00945_bloom_filter_index` where you read 6 rows from the table (based on the bloom filter) + 1 in the subquery (to generate `(SELECT arrayJoin([toInt32(1), toInt32(2)]))`). I understand that the desired behaviour in both cases would be to count all rows in the IN filter and the test need fixing, right?

* There might be other issues still regarding duplicated progress now that all pipelines inside the query should be reporting.

Algunenano committed:
Rework how progress is reported in views

Algunenano commented:
Some extra issues found from this change:
* The materialization/live/window views should be improved and uniformized. For example, the original query had only its `read_rows` but the global (with mat views) `written_rows`. That should be uniformized.
* `GLOBAL IN` reads more rows than `IN` (xp_d is distributed)
   - `select * from xp_d where i in (select * from numbers(2));` reads 4 rows (2 from numbers, 2 from table)
   - `select * from xp_d where i in (select * from numbers(2));` reads 6 rows (2 from numbers, 2 from table, 2 I don't know where yet)

Algunenano committed:
Adapt tests to count properly

Algunenano committed:
Style

Algunenano committed:
Avoid flakyness due to print order

Algunenano committed:
Adapt 01504_rocksdb

Algunenano committed:
Mark affected tests as short

Algunenano commented:
It's not representative of performance (since the query is doing nothing) but you can see how many things start to work better / more consistently (killing queries, limits, progress reports) with this PR:

![image](https://user-images.githubusercontent.com/664253/145551562-507a8177-7917-466a-8b76-85524f0a158f.png)

Algunenano commented:
Tests:
* FunctionalStatelessTestTsan: Seems like the runner died. It seems to have happened in other PRs too randomly but there aren't any (visible to me) logs so although I think it's unrelated I don't have proof.
* Integration tests (both asan, release and thread): Flaky tests (test_prometheus_endpoint -> https://github.com/ClickHouse/ClickHouse/pull/32476)

Everything else looks clean.

KochetovNicolai commented:
ok, it looks like perf tests are faster now because `max_rows_to_read` took a subquery into account.
I think it's a backward incompatibility, and we'd better keep the old behaviour.

Algunenano commented:
> ok, it looks like perf tests are faster now because max_rows_to_read took a subquery into account.
I think it's a backward incompatibility, and we'd better keep the old behaviour.

I'm not sure about that. I agree that it is a backward incompatible change but I'd say it's a desirable one since you are reading from a different table in a subquery and I don't think `IN` subquery should be any special. Otherwise, can we define what counts as a `read_row` and what doesn't?

I'm not saying this PR's behaviour is the best, but ignoring read rows from the `IN` subquery is unexpected and leads to confusing metrics. I really want to know if the subquery is reading or trying to read 10B rows to prepare the filter.

Algunenano commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'master' into kill_scalar_github

Algunenano commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'master' into kill_scalar_github

Algunenano commented:
@KochetovNicolai what do you think is needed to be able to move this forward? If you let me know your concerns about the changes I'll work to polish them and get the PR into a better shape.

I want to work on recovering the performance impact that https://github.com/ClickHouse/ClickHouse/pull/32811 has on materialized views with scalar queries (after that PR they aren't cached to avoid issues when it uses the target table) and having this merged would be useful so I can actually test that the scalar subqueries are/aren't executed.

KochetovNicolai commented:
@Algunenano sorry for not looking into this pr again.

> I agree that it is a backward incompatible change but I'd say it's a desirable one since you are reading from a different table in a subquery and I don't think IN subquery should be any special.

I also agree that such a logic may be more reasonable and desirable. At the same time, I feel like it's a very bad idea to brake such a compatibility. Some previously working queries will start to throw exceptions after update, and I could not imagine how I should I convince people  to fix their production instead of our code.

There are a few possible things which we can do:
* Keep ignoring `max_rows_to_read` for subqueries and discuss possible changes in a separate pr. It's the easiest one, I would be happy to merge this pr. (There could be other settings which are not obvious, and we should be ready to default them too).
* Mark this pr as `backward incompatibly`, and introduce another one setting to roll back to old behaviour (which we can remove probably in a few month). (I don't like new settings).
* Also add a new setting, but keep current behaviour as default.
* Separate settings for subqueries (which may be zero or same for query by default).

Algunenano commented:
Ok, I understand.
To me the simplest way forward would be to ignore rows read in scalar queries and create a ticket to review what should be the behaviour. I'll see what can be done to implement now (now the code is clean since it doesn't need to know which context is checking limits).

I understand that controlling scalar subqueries is even in the 2022 roadmap (`Make scalar subqueries accountable.`) so a ticket would give other people the opportunity to comment on it. OTOH, 22.1 seems like a good moment to break compatibility ;D

Algunenano committed:
Merge remote-tracking branch 'blessed/master' into kill_scalar_github

Algunenano commented:
After looking into it for a while adding the exception while keeping the hability to kill / cancel scalar queries requires quite a bit of awful code. We are using `QueryStatus` to check progress, limits and query cancellation so to have one an not the others we would need to either create a new `QueryStatus` that shares only part of the state of the class (the kill flag and the executors but not the progress) or come up with a different way of handling them. It's doable but I don't see an easy way of just adding a few lines of code that can be removed in the future, and I see a "I need to have subclasses for `QueryStatus` which I would need to share via mutexes just to handle this".

Let's open a ticket and discuss it first before spending hours working and testing a workaround that might not be needed if we decide to declare the change `backwards incompatible` and count all the rows in the same way.

Algunenano commented:
Ticket here: https://github.com/ClickHouse/ClickHouse/issues/32993

Algunenano committed:
Merge remote-tracking branch 'blessed/master' into kill_scalar_github

Algunenano committed:
Move tests

Algunenano committed:
Add some basic tests to read rows with scalar subqueries

Algunenano commented:
It seems like a breaking change with a good message is the way to go, so I'm adding some extra tests around this change and will edit the message to make it clear it's not only about KILL anymore and how things are accounted for (cached when possible and so on).

Algunenano commented:
It seems that the HTTP Handler is still not getting the progress of scalar subqueries and reports less rows:

```
curl 'http://localhost:8123/?add_http_cors_header=1&default_format=JSONCompact&max_result_rows=1000&max_result_bytes=10000000&result_overflow_mode=break&wait_end_of_query=1'   -H 'Connection: keep-alive'   -H 'Pragma: no-cache'   -H 'Cache-Control: no-cache'   -H 'sec-ch-ua: " Not A;Brand";v="99", "Chromium";v="96"'   -H 'DNT: 1'   -H 'sec-ch-ua-mobile: ?0'   -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36'   -H 'sec-ch-ua-platform: "Linux"'   -H 'Content-Type: text/plain;charset=UTF-8'   -H 'Accept: */*'   -H 'Origin: https://play-ci.clickhouse.com'   -H 'Sec-Fetch-Site: cross-site'   -H 'Sec-Fetch-Mode: cors'   -H 'Sec-Fetch-Dest: empty'   -H 'Accept-Language: en-GB,en;q=0.9,es;q=0.8'   --data-raw $'SELECT \'#02136_scalar_subquery_1\', (SELECT max(number) FROM numbers(1000)) as n;'   --compressed -vvv
*   Trying 127.0.0.1:8123...
* Connected to localhost (127.0.0.1) port 8123 (#0)
> POST /?add_http_cors_header=1&default_format=JSONCompact&max_result_rows=1000&max_result_bytes=10000000&result_overflow_mode=break&wait_end_of_query=1 HTTP/1.1
> Host: localhost:8123
> Accept-Encoding: deflate, gzip, br, zstd
> Connection: keep-alive
> Pragma: no-cache
> Cache-Control: no-cache
> sec-ch-ua: " Not A;Brand";v="99", "Chromium";v="96"
> DNT: 1
> sec-ch-ua-mobile: ?0
> User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36
> sec-ch-ua-platform: "Linux"
> Content-Type: text/plain;charset=UTF-8
> Accept: */*
> Origin: https://play-ci.clickhouse.com
> Sec-Fetch-Site: cross-site
> Sec-Fetch-Mode: cors
> Sec-Fetch-Dest: empty
> Accept-Language: en-GB,en;q=0.9,es;q=0.8
> Content-Length: 80
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Wed, 22 Dec 2021 18:12:08 GMT
< Connection: Keep-Alive
< Content-Type: application/json; charset=UTF-8
< X-ClickHouse-Server-Display-Name: Mordor
< Transfer-Encoding: chunked
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Headers: origin, x-requested-with
< Access-Control-Allow-Methods: POST, GET, OPTIONS
< Access-Control-Max-Age: 86400
< X-ClickHouse-Query-Id: 3a977e51-a801-4b40-8be6-0a1f9ba894cf
< X-ClickHouse-Format: JSONCompact
< X-ClickHouse-Timezone: Europe/Madrid
< Keep-Alive: timeout=3
< X-ClickHouse-Summary: {"read_rows":"0","read_bytes":"0","written_rows":"0","written_bytes":"0","total_rows_to_read":"0"}
< 
{
        "meta":
        [
                {
                        "name": "'#02136_scalar_subquery_1'",
                        "type": "String"
                },
                {
                        "name": "n",
                        "type": "Nullable(UInt64)"
                }
        ],

        "data":
        [
                ["#02136_scalar_subquery_1", "999"]
        ],

        "rows": 1,

        "statistics":
        {
                "elapsed": 0.000292506,
                "rows_read": 1,
                "bytes_read": 1
        }
}
```

I'll have a look

Algunenano committed:
Hook progress callbacks with subqueries too

Algunenano committed:
Add existing progress to the record of the output format progress

KochetovNicolai commented on tests/performance/set_index.xml:
```diff
@@ -3,17 +3,17 @@
     <!-- IN is used at index analysis -->
-    <query>SELECT count() FROM test_in WHERE a IN (SELECT rand(1) FROM numbers(200000)) SETTINGS max_rows_to_read = 1, read_overflow_mode = 'break'</query>
+    <query short="1">SELECT count() FROM test_in WHERE a IN (SELECT rand(1) FROM numbers(200000)) SETTINGS max_rows_to_read = 1, read_overflow_mode = 'break'</query>
 
-    <query>SELECT count() FROM test_in WHERE toInt64(a) IN (SELECT toInt64(rand(1)) FROM numbers(200000)) settings max_rows_to_read=1, read_overflow_mode='break'</query>
+    <query short="1">SELECT count() FROM test_in WHERE toInt64(a) IN (SELECT toInt64(rand(1)) FROM numbers(200000)) settings max_rows_to_read=1, read_overflow_mode='break'</query>
```
This tests were checking the performance of SET creation. We need to fix it a little bit.

Algunenano replied:
I could switch it to max_rows_to_read=200000. That way it will still overflow but it will do it after executing the subquery.

Algunenano committed:
Change perf set_index test to check build index

Algunenano committed:
Merge remote-tracking branch 'blessed/master' into kill_scalar_github

Algunenano committed:
Test that JSON output statistics include the rows read in scalar subqueries

Algunenano committed:
Check that progress is sent while processing scalar subqueries

Algunenano commented:
* Hooked the progress of subqueries too. Covers the summary too.
* Fixed the progress of JSON statistics with scalar subqueries: since the scalar subquery has been executed before creating the output format it needs to be populated with already read statistics.
* Changed perf tests to read +1 row than the subquery so the index should e generated.
